### PR TITLE
Initial commit for bart.

### PR DESCRIPTION
**Pull request by trivialfis**
Introduced in #4398, this is a draft implementation of  **[Bayesian Additive Regression Tree](https://arxiv.org/pdf/0806.3286.pdf)**.  It's a tree ensemble model with MCMC(CGM98 gibbs) as structure learning algorithm.  Although  XGBoost is a gradient boosting library, which is quite different than sampling methods.  There are Bayesian methods that use gradient, but to my knowledge non has apply it on tree model yet.  I'm interested in this subject and want to work on it in the future if possible.  I hope that implementing bart can open up the possibility of Bayesian view in XGBoost.

Currently it's far from usable, notably missing a predictor, classification support, burn in support and optimization is non-exist.  Also, please ignore the styling.

Other than these, there are a few design issues to be addressed:
- gibbs sampling doesn't use gradient at all so there will be conflict with current internal structure.
- sampling based method is kind of hard to test.
- missing data handling is difficult since we can't just accumulate some statistic.

In this PR a `GibbsUpdater` is created to coordinate the sampling iteration, while `TreeMutation` class is an enhanced version of `RegTree` with specific sampling features.  New model lives in plugin for now.  I would like to know the opinion from community about this added feature. The development will continue if it's accepted as an useful model and suitable to live in XGBoost.

@tqchen @hcho3 @RAMitchell @CodingCat and anyone who's interested. ;-)

### CODE REVIEW

**Comment by CodingCat** [comment]
It may take a while for me to read the paper first...

**Comment by CodingCat** [comment]
@yinlou looks like relevant to your research

**Comment by yinlou** [comment]
Overall structure makes sense to me. It would be interesting to include some benchmark results on open datasets to show the benefits of BART.

**Comment by trivialfis** [comment]
I have a local implementation that is more complete and testable.  By now I have some experience with various Bayesian methods.  I think one of the problems of these models that rarely mentioned in literatures, is their heavy usage of `log` and `exp`.  It's very difficult to avoid numeric overflow/underflow.

**Comment by trivialfis** [comment]
Nah, let's stay with GB.  For anyone interested in where did this go, see https://github.com/trivialfis/kapok .

**Comment by trivialfis** [comment]
Here's a simple explanation why I dropped the idea of merging it into XGBoost, for anyone who's interested.  And sorry for the noise.

BART, like many other ensemble model, one first train a weak model, ask it for prediction, then subtract its prediction from true labels (y) and get the error.  After that we train another weak model, use previous error as the new target, add the second model's prediction to the first one to reduce total error, and so on.  Instead of using greedy exploration, BART uses sampling method, with each stage's error combined with other factors as the sampling mean, specifically mean of Gaussian distribution for leaves' value.

During each stage, it draws leaves' value non-deterministically, which means it can draw a good tree that reduces the error.  But it's also possible that it draw a bad tree that somehow enlarge the error.  Mathematically the error should still converge to zero.  But in my experiments, due to its heavy use of 'log', 'exp' and division, numeric overflows happens easily.  If in one of the training iterations we get a numeric overflow for one leaf (we prefer small trees so every leaf matters), all the previous training is basically ruined.  And the next iteration will use the mean containing this ridiculous error to sample new values, therefore it won't be corrected quickly.  In general it's not a very robust way to train a model.

A possible fix at the algorithm level is, we somehow add rejection sampling for leaves' value and related factors, that accepts drawn values probabilistically conditioned on new errors.  But that would be a research project for the future. ;-)  Thanks for the attention on this.

@yinlou @RAMitchell @CodingCat @hcho3
